rclcpp/include/rclcpp/context.hpp

```diff
@@ -47,6 +48,15 @@ class ContextAlreadyInitialized : public std::runtime_error
 /// Forward declare WeakContextsWrapper
 class WeakContextsWrapper;
 
+struct OnShutdownCallbackHandle
+{
+  RCLCPP_SMART_PTR_DEFINITIONS(OnShutdownCallbackHandle)
+
+  using OnShutdownCallbackType = std::function<void ()>;
+
+  OnShutdownCallbackType callback;
+};
+
```

Comment:
To avoid leaking implementation details, I would define `OnShutdownCallback`/`OnShutdownCallbackHandle` like:

```cpp
using OnShutdownCallbackType = std::function<void ()>;
class OnShutdownCallbackHandle
{
  friend rclcpp::Context;
  std::weak_ptr<OnShutdownCallbackType> handle_;
};
```

So in the future `OnShutdownCallbackHandle` can be something different than a std::weak_ptr.
They can be defined nested in `rclcpp::Context` or directly in the `rclcpp` namespace, I don't mind strongly.

Reply:
Thanks.  
But I think we cannot use `std::weak_ptr<OnShutdownCallbackType> handle_;`. We should hold handle in Context.

Reply:
The `Context` container for the callbacks is `std::unordered_set<std::shared_ptr<OnShutdownCallbackType>>`, in the same fashion than now.

You can seamlessly convert between `OnShutdownCallbackHandle` and `std::shared_ptr<OnShutdownCallbackType>` internally.

e.g.
```cpp
OnShutdownCallbackHandle
on_shutdown(OnShutdownCallbackType callback) {
  auto callback_shared_ptr = std::make_shared<OnShutdownCallbackType>(callback);
  on_shutdown_callbacks_.emplace(callback_shared_ptr);
  OnShutdownCallbackHandle handle;
  handle.handle_ = callback_shared_ptr;
  return handle;
}

bool
remove_on_shutdown_callback(OnShutdownCallbackHandle handle) {
  auto callback_shared_ptr = handle.handle_.lock();
  if (callback_shared_ptr == nullptr) {
    return false;
  }
  return on_shutdown_callbacks_.erase(callback_shared_ptr) == 1;
}
```

Reply:
Sorry, I think I didn't explain my idea correctly.

The point of my comment was to make the return type **exactly** `OnShutdownCallbackHandle` and not `OnShutdownCallbackHandle::WeakPtr`, so the fact that we're using a `std::weak_ptr` as a handle and internally storing a `shared_ptr` is completly transparet for the user.
In the future if we use another approach and `OnShutdownCallbackHandle` is an integer (or whatever other thing), the API doesn't change and we only break ABI.

Reply:
Thanks for further explanation.  
Get it.  
I will update codes.